resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/kubernetes/KubernetesExternalShuffleService.scala

```diff
@@ -0,0 +1,133 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.deploy.kubernetes
+
+import java.nio.ByteBuffer
+
+import scala.collection.JavaConverters._
+
+import io.fabric8.kubernetes.api.model.Pod
+import io.fabric8.kubernetes.client.{KubernetesClientException, Watch, Watcher}
+import io.fabric8.kubernetes.client.Watcher.Action
+
+import org.apache.spark.{SecurityManager, SparkConf}
+import org.apache.spark.deploy.ExternalShuffleService
+import org.apache.spark.deploy.kubernetes.constants._
+import org.apache.spark.internal.Logging
+import org.apache.spark.network.client.{RpcResponseCallback, TransportClient}
+import org.apache.spark.network.shuffle.ExternalShuffleBlockHandler
+import org.apache.spark.network.shuffle.protocol.BlockTransferMessage
+import org.apache.spark.network.shuffle.protocol.kubernetes.RegisterKubernetesApp
+import org.apache.spark.network.util.TransportConf
+import org.apache.spark.scheduler.cluster.kubernetes.DriverPodKubernetesClientProvider
+
+
+/**
+ * An RPC endpoint that receives registration requests from Spark drivers running on Kubernetes.
+ * It detects driver termination and calls the cleanup callback to [[ExternalShuffleService]].
+ */
+private[spark] class KubernetesShuffleBlockHandler (
+  sparkConf: SparkConf,
+  transportConf: TransportConf)
+  extends ExternalShuffleBlockHandler(transportConf, null) with Logging {
+
+  private var connectedApps: Set[String] = Set()
+  private val CONNECTED_APPS_LOCK = new Object
+  private val Watch: Option[Watch] = ShuffleWatcher()
+
+  protected override def handleMessage(
+    message: BlockTransferMessage,
+    client: TransportClient,
+    callback: RpcResponseCallback): Unit = {
+      message match {
+        case RegisterDriverParam(appId) =>
+          val address = client.getSocketAddress
+          logInfo(s"Received registration request from app $appId (remote address $address).")
+          CONNECTED_APPS_LOCK.synchronized {
+            if (connectedApps.contains(appId)) {
+              logWarning(s"Received a registration request from app $appId, but it was already " +
+                s"registered")
+            }
+            connectedApps += appId
+          }
+          callback.onSuccess(ByteBuffer.allocate(0))
+        case _ => super.handleMessage(message, client, callback)
+      }
+  }
+
+  private def ShuffleWatcher(): Option[Watch] = {
+    try {
+      val kubernetesClient = new DriverPodKubernetesClientProvider(sparkConf, "default").get
+      Some(kubernetesClient
+        .pods()
+        .withLabels(Map(SPARK_ROLE_LABEL -> "driver").asJava)
+        .watch(new Watcher[Pod] {
+          override def eventReceived(action: Watcher.Action, p: Pod): Unit = {
+            action match {
+              case Action.DELETED | Action.ERROR =>
+                val appId = p.getMetadata.getLabels.getOrDefault(SPARK_APP_ID_LABEL, "")
```

Comment:
We probably don't want to do anything at all if the app ID label is missing. There should be no need to set a default value for the label in that case.